File: communication/src/main/java/net/tomofiles/skysign/communication/domain/vehicle/Vehicle.java
package net.tomofiles.skysign.communication.domain.vehicle;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import net.tomofiles.skysign.communication.domain.communication.CommunicationId;
import net.tomofiles.skysign.communication.event.EmptyPublisher;
import net.tomofiles.skysign.communication.event.Publisher;

@EqualsAndHashCode(of = {"id"})
@ToString
public class Vehicle {
    @Getter
    private final VehicleId id;

    private final Generator generator;
    
    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private String vehicleName = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private CommunicationId commId = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version version;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version newVersion;

    @Setter
    private Publisher publisher = new EmptyPublisher();
    
    Vehicle(VehicleId id, Version version, Generator generator) {
        this.id = id;
        this.version = version;
        this.newVersion = version;

        this.generator = generator;
    }

    public void nameVehicle(String name) {
        this.vehicleName = name;
        this.newVersion = this.generator.newVersion();
    }

    public void giveCommId(CommunicationId id) {
        CommunicationId beforeId = this.commId;
        this.commId = id;
        this.newVersion = this.generator.newVersion();
        this.publisher
                .publish(
                        new CommunicationIdChangedEvent(beforeId, id, this.id, this.newVersion));
    }
}

File: fleet-formation/pkg/vehicle/domain/vehicle/vehicle.go
package vehicle

import (
	"errors"
	"fleet-formation/pkg/vehicle/domain/event"
)

// ID .
type ID string

// CommunicationID .
type CommunicationID string

// Version .
type Version string

const (
	// Original .
	Original = false
	// CarbonCopy .
	CarbonCopy = true
)

var (
	// ErrCannotChange .
	ErrCannotChange = errors.New("cannot change carbon copied vehicle")
)

// Vehicle .
type Vehicle struct {
	id              ID
	name            string
	communicationID CommunicationID
	isCarbonCopy    bool
	version         Version
	newVersion      Version
	gen             Generator
	pub             event.Publisher
}

// SetPublisher .
func (v *Vehicle) SetPublisher(pub event.Publisher) {
	v.pub = pub
}

// GetID .
func (v *Vehicle) GetID() ID {
	return v.id
}

// GetName .
func (v *Vehicle) GetName() string {
	return v.name
}

// GetCommunicationID .
func (v *Vehicle) GetCommunicationID() CommunicationID {
	return v.communicationID
}

// GetVersion .
func (v *Vehicle) GetVersion() Version {
	return v.version
}

// GetNewVersion .
func (v *Vehicle) GetNewVersion() Version {
	return v.newVersion
}

// NameVehicle .
func (v *Vehicle) NameVehicle(name string) error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	v.name = name
	v.newVersion = v.gen.NewVersion()
	return nil
}

// GiveCommunication .
func (v *Vehicle) GiveCommunication(communicationID CommunicationID) error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	if v.pub != nil {
		if v.communicationID == "" {
			v.communicationID = communicationID
			v.newVersion = v.gen.NewVersion()
			v.pub.Publish(CommunicationIDGaveEvent{
				CommunicationID: v.communicationID,
			})
		} else {
			beforeId := v.communicationID
			v.communicationID = communicationID
			v.newVersion = v.gen.NewVersion()
			v.pub.Publish(CommunicationIDRemovedEvent{
				CommunicationID: beforeId,
			})
			v.pub.Publish(CommunicationIDGaveEvent{
				CommunicationID: v.communicationID,
			})
		}
	}
	return nil
}

// RemoveCommunication .
func (v *Vehicle) RemoveCommunication() error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	if v.pub != nil {
		v.pub.Publish(CommunicationIDRemovedEvent{
			CommunicationID: v.communicationID,
		})
	}
	v.communicationID = ""
	v.newVersion = v.gen.NewVersion()
	return nil
}

// Generator .
type Generator interface {
	NewID() ID
	NewVersion() Version
}
